### PR TITLE
Initial earcut.hpp integration

### PR DESCRIPTION
**Pull request by randypalusz**
Tracking #107 - Maybe too early for a PR. Added earcut.hpp along with some tests. Added a skeleton for the upcoming SolidPolygonLayer example. Currently builds properly, but nothing is rendering yet because tesselation/shaders aren't implemented.

### CODE REVIEW

File: cpp/modules/deck.gl/core/src/lib/earcut.hpp [hunk 1]
```diff
@@ -0,0 +1,823 @@
+// ISC License
```
Comment (ilijapuaca) [inline cpp/modules/deck.gl/core/src/lib/earcut.hpp:1]: Perhaps we should try to incorporate this in our dependency repo somehow, mainly in order to be able to do version control in a more convenient way.

This should be good for now though

File: cpp/examples/deck.gl/vancouver-blocks.cc [hunk 2]
```diff
@@ -0,0 +1,79 @@
...
+  props->getPolygon = [](const Row &row) { return row.getVector3<float>("coordinates"); };
+
+  return props;
+}
+
+int main(int argc, const char *argv[]) {
+  // Get data file paths relative to working directory
+  auto programPath = std::string{argv[0]};
+  auto programDirectory = programPath.erase(programPath.find_last_of("/"));
+  auto vancouverDataPath = programDirectory + "/data/vancouver-blocks.ndjson";
```
Comment (ilijapuaca) [inline cpp/examples/deck.gl/vancouver-blocks.cc:60]: This doesn't seem to be used?

File: cpp/modules/deck.gl/core/test/lib/earcut-test.cpp [hunk 3]
```diff
@@ -0,0 +1,75 @@
...
+
+  polygon.push_back(mainPolygon);
+  polygon.push_back(holes);
+
+  std::vector<N> indices = mapbox::earcut<N>(polygon);
+
+  // Length of indices list should be 3* that of the input polygon list
+  EXPECT_EQ(indices.size() / 3, (mainPolygon.size() + holes.size()));
+
+  //   std::cout << "Indices: ";
```
Comment (ilijapuaca) [inline cpp/modules/deck.gl/core/test/lib/earcut-test.cpp:46]: Remove commented code

File: cpp/modules/deck.gl/core/test/lib/earcut-test.cpp [hunk 4]
```diff
@@ -0,0 +1,75 @@
...
+
+  polygon.push_back(mainPolygon);
+  polygon.push_back(holes);
+
+  std::vector<N> indices = mapbox::earcut<N>(polygon);
+
+  // Length of indices list should be 3* that of the input polygon list
+  EXPECT_EQ(indices.size() / 3, (mainPolygon.size() + holes.size()));
+
+  //   std::cout << "Indices3: ";
```
Comment (ilijapuaca) [inline cpp/modules/deck.gl/core/test/lib/earcut-test.cpp:70]: Remove commented code

File: cpp/modules/deck.gl/layers/src/solid-polygon-layer/solid-polygon-layer.cpp [hunk 5]
```diff
@@ -66,7 +71,9 @@ auto SolidPolygonLayer::Props::getProperties() const -> const Properties* {
 void SolidPolygonLayer::initializeState() {
   // Using lambdas over std::bind - potential C++ retain cycle issue
 
-  auto polygon = std::make_shared<arrow::Field>("instancePolygons", arrow::fixed_size_list(arrow::float32(), 4));
+  // list<fixed_size_list(Vector3)>
```
Comment (ilijapuaca) [inline cpp/modules/deck.gl/layers/src/solid-polygon-layer/solid-polygon-layer.cpp:74]: Remove commented code

File: cpp/examples/deck.gl/vancouver-blocks.cc (hunk 2 above)
Comment (randypalusz) [reply cpp/examples/deck.gl/vancouver-blocks.cc:60]: Looking at the JS example [here](https://github.com/visgl/deck.gl/blob/8.1-release/examples/website/geojson/app.js), the vancouver-blocks data is being imported into a new GeoJson layer, and the land-cover data was simply a const list of coordinates. Not sure if this vancouver-blocks data can be incorporated into an example just using the SolidPolygonLayer, but it's here as a placeholder for now.

File: cpp/modules/deck.gl/core/src/lib/earcut.hpp (hunk 1 above)
Comment (randypalusz) [reply cpp/modules/deck.gl/core/src/lib/earcut.hpp:1]: Talking with Ib, it was suggested to just copy/paste into deck.gl along with the copyright message at the top of course. Definitely makes sense to later incorporate this into the deps repo.

File: cpp/examples/deck.gl/vancouver-blocks.cc (hunk 2 above)
Comment (ibgreen) [reply cpp/examples/deck.gl/vancouver-blocks.cc:60]: So geojson polygons have a variable number of coordinates per polygon and will need to be handled using variable length Arrow lists, and we need to extract each polygon outlines's vertices and pass them to earcut, and then concatenate them (or just SubData the resulting tesselated polygons) back into the GPU buffer.

@ilijapuaca you will probably need to give Randy a few pointers here.

File: cpp/modules/deck.gl/core/src/lib/earcut.hpp (hunk 1 above)
Comment (ibgreen) [reply cpp/modules/deck.gl/core/src/lib/earcut.hpp:1]: Yes this is fine. You can check the repo but I strongly doubt that there is much activity in this file, it is not like a dependency that we need to keep fresh. We could probably just treat it as a forked file.

**Pull request merged**